Logica del juego

cvillar97 opened the pull request.

cvillar97 committed:
Crea maquetado de web

El cliente solicito una web app del juego Simón Dice.

cvillar97 committed:
Agrega colores

El cliente solicitó el uso de un circulo dividido entre los 4 colores para hacer el juego.

cvillar97 committed:
Agrega estilos no definitivos

Se agregaron estilos a la web para trabajar de forma mas cómoda en el desarrollo del juego.

cvillar97 committed:
Modifica estructura HTML

Se agrego un nuevo label donde se pondrá el numero de la ronda correspondiente con JS.
Se borraron los ID's de los colores y se les asigno una misma class "colores" para luego facilitar el desarrollo de la lógica del juego en JS.

cvillar97 committed:
Crea archivo JS

Se creó el archivo main.js donde se desarrollara la lógica que necesita el juego.

cvillar97 committed:
Crea contenedores y contador de ronda

En estos arrays se guardaran las secuencias de juego de la maquina y el usuario respectivamente.

cvillar97 committed:
Guarda botón Jugar en variable

Al hacer click en el botón se ejecutará el código que dará inicio a la secuencia del juego.

cvillar97 committed:
Crea función actualizarNumeroRonda

Esta función va a actualizar el numero de ronda cada vez que se inicie una nueva.

cvillar97 committed:
Crea función actualizarEstadoDeJuego

La función va a actualizar el estado a Turno máquina o Turno jugador.

cvillar97 committed:
Agrega espacio al final del código

cvillar97 committed:
Crea función obtenerNumeroRandom

Sera usada dentro de la función obtenerColorRandom para generar un numero entero aleatorio entre el 0 y 3.

cvillar97 committed:
Crea función obtenerColorRandom

Devuelve de forma aleatoria 1 de los 4 colores del circulo.

cvillar97 committed:
Crea función bloquearInputUsuario

Servirá para impedir que el usuario seleccione nuevos colores cuando sea el turno de la máquina.

cvillar97 committed:
Crea función reiniciarJuego

Llevará toda la interfaz del usuario, secuencias y ronda a los valores iniciales cuando sea llamada en caso de perder o al presionar el botón "Jugar".

cvillar97 committed:
Agrega función reiniciarJuego

Lleva la interfaz a valores por default al comenzar el juego.

cvillar97 committed:
Trabaja en función manejarRonda

Dicha función ejecuta la secuencia de eventos que ocurrirán durante cada ronda. Todavía no estaría terminada.

cvillar97 committed:
Agrega numeral faltante

Corrige error.

cvillar97 committed:
Crea variable nuevoColor

En la que se guardará cada nuevo color obtenido en cada ronda.

cvillar97 committed:
Corrige errores en 2 funciones

Estaba asignándole un nuevo valor a las variables en lugar de modificar el textContent de las mismas.

cvillar97 committed:
Agrega estilo a colores del juego

Se les asignó una opacidad de 0.5 para que luego pueda llevarse la opacidad a 1, generando un efecto de iluminación. Agregué la propiedad transition para que la opacidad sea modificada gradualmente en 500ms cada vez que haga falta.

cvillar97 committed:
LLama a manejarRonda() dentro de onclick

cvillar97 committed:
Crea función resaltarColor

Sera la encargada de modificar la opacidad de los colores para que estos se iluminen y puedan ser identificados por el usuario. Para esto se agrego un setTimeout que en 500ms le cambia la opacidad al valor inicial de 0.5. Cada iluminación dura 1s en total.

cvillar97 committed:
Guarda colores y resalta cada uno

Se guardan los colores en el array que es recorrido por un forEach que resalta cada color. Todavía no esta terminado.

cvillar97 committed:
Indenta código con ALT + SHIFT + F

cvillar97 committed:
Crea retrasos de tiempo

Se crearon retrasos en milisegundos para que la máquina pueda terminar de jugar(RETRASO_MS_MAQUINA) y para que se muestre cada color al usuario(RETRASO_MS_COLOR).

cvillar97 committed:
Crea setTimeout

Este setTimeout servirá para que una vez que termine de jugar la máquina se ejecuten los eventos necesarios para que el usuario comience a jugar.

cvillar97 committed:
Crea función desbloquearInputUsuario

Servirá para que el usuario pueda comenzar a jugar haciendo click sobre los colores llamando en ese momento a la función manejarInputUsuario.

cvillar97 committed:
Crea función manejarInputUsuario

Será la encargada de hacer una comparación entre los colores de la máquina y los colores elegidos por el usuario para decidir si continua a la siguiente ronda o pierde.
En la variable $color guardo el color seleccionado por el usuario gracias al parámetro "e" que es un objeto que contiene entre muchas la key "target" cuyo valor es el color al que le hizo click el usuario.

cvillar97 committed:
Guarda color y comienza la comparación

Guardamos el $color en secuenciaUsuario y definimos $colorMaquina para que cada vez que el usuario haga clic se compare el mismo índice en las 2 secuencias.
El primer condicional sería el caso de éxito donde se continúa jugando.

cvillar97 committed:
Indenta con ALT + SHIFT + F

cvillar97 committed:
Agrega nueva condición

Este es el caso de que el color seleccionado sea el correcto pero todavía no se hayan comparado todos los colores de secuencia maquina.

cvillar97 committed:
Agrega función faltante

Al hacer clic sobre los colores no seguía el juego.

cvillar97 committed:
Borra ( )

No es necesario para llamar a la función.

cvillar97 committed:
Agrega función resaltarColor

Para resaltar el color elegido por el usuario.

cvillar97 committed:
Agrega caso de juego

En caso de que el color seleccionado sea correcto y además se hayan comparado todos los colores de secuencia máquina con los seleccionados por el usuario, se ejecutará una nueva ronda y la secuencia del usuario se reseteara.

cvillar97 committed:
Agrega caso de color incorrecto

En caso de que el color elegido sea diferente del resaltado por la máquina se reinicia el juego.

cvillar97 committed:
Corrige función bloquearInputUsuario

Debía recorrer cada elemento del node list con un bucle.

cvillar97 committed:
Corrige función bloquear Input Usuario

Faltaba agregar el nombre de los elementos contenidos en el Node list antes del .onclick para que pueda funcionar correctamente.

fsodano commented on main.js:
```diff
@@ -0,0 +1,112 @@
+    actualizarNumeroRonda()
+    actualizarEstadoDeJuego("Turno máquina")
+    bloquearInputUsuario()
+    let nuevoColor = obtenerColorRandom()
```
const por default, siempre

fsodano commented on main.js:
```diff
@@ -0,0 +1,112 @@
+    }, RETRASO_MS_MAQUINA)
+
+    secuenciaMaquina.forEach(function (color, index) {
+        let RETRASO_MS_COLOR = (index + 1) * 1000
```
RETRASO_MS_RESTALTADO_COLOR

fsodano commented on main.js:
```diff
@@ -0,0 +1,112 @@
+}
+
+function obtenerNumeroRandom() {
+    let numeroRandom = Math.floor(Math.random() * 4)
```
este 4 es un número mágico, en vez de hacer eso guardate en algún lado el array de colores y hacés colores.length

cvillar97 commented on index.html:
```diff
@@ -0,0 +1,31 @@
+<body>
+    <section id="estado-de-juego">
+        <label>Ronda #</label><label id="numero-ronda" >-</label>
+        <label id="turno">Turno -</label>
```
@fsodano muchas gracias por la review! No entendí este mensaje

fsodano commented on index.html:
```diff
@@ -0,0 +1,31 @@
+<body>
+    <section id="estado-de-juego">
+        <label>Ronda #</label><label id="numero-ronda" >-</label>
+        <label id="turno">Turno -</label>
```
Quedó colgado, ignoralo.

cvillar97 committed:
Instala prettier en VS Code

Se instaló complemento Prettier en el IDE para mejorar la identacion del código de forma automática.

cvillar97 committed:
Agrega variable para colores

Contendrá los colores del juego.

cvillar97 committed:
Reemplaza magic number

Se reemplazó el magic number "4" por colores.length para comprender mejor el código.

cvillar97 committed:
Corrige nombre de variable

Expresa mejor la función de la misma.

cvillar97 committed:
Corrige tipo de variable

En este caso es mejor el uso de const por default.
